README.md
rOCAT
================

-   [1 Installation](#1-installation)
-   [2 Example](#2-example)
    -   [2.1 Working with Single
        Dataset](#21-working-with-single-dataset)
        -   [2.1.1 load package](#211-load-package)
        -   [2.1.2 import data](#212-import-data)
        -   [2.1.3 run OCAT pipeline](#213-run-ocat-pipeline)
        -   [2.1.4 clustering and
            evaluation](#214-clustering-and-evaluation)
    -   [2.2 Working with Multiple
        Datasets](#22-working-with-multiple-datasets)
        -   [2.2.1 import datasets](#221-import-datasets)
        -   [2.2.2 run OCAT pipline](#222-run-ocat-pipline)
        -   [2.2.3 clustering and
            evaluation](#223-clustering-and-evaluation)

<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- badges: start -->

[![Travis build
status](https://travis-ci.com/bowang-lab/rOCAT.svg?branch=main)](https://travis-ci.com/bowang-lab/rOCAT)

<!-- badges: end -->

This is the R package for One Cell At A Time(OCAT), which provides a
fast and memory-efficient framework for analyzing and integrating
large-scale scRNA-seq data. Details of the method can be check
[here](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02659-1).

# 1 Installation

You can install the development version of rOCAT from
[GitHub](https://github.com/bowang-lab/rOCAT) with:

``` r
# install.packages("devtools")
devtools::install_github("bowang-lab/rOCAT")
```

# 2 Example

Here we demonstrate how to sparsely encodes single-cell gene expression
data on both single and multiple datasets

The **single** dataset using 3,005 cells and 4,412 genes in the mouse
somatosensory cortex and hippocampal CA1 region from Zeisel et
al. (2015) and can be download
[here](https://github.com/bowang-lab/OCAT/blob/master/vignettes/Clustering/Test_5_Zeisel.mat).

The **multiple** datasets consist of five scRNA-seq datasets (Baron et
al. 2016, Muraro et al. 2016, Segerstolpe et al. 2016, Wang et al. 2016,
Xin et al. 2016) and can be download
[here](https://data.wanglab.ml/OCAT/Pancreas.zip).

## 2.1 Working with Single Dataset

### 2.1.1 load package

``` r
library(rOCAT)
library(reticulate) # must load this to config the python env

# if no python environment exists in the machine
reticulate::py_config()
# reticulate::install_miniconda() # should prompt to ask for installation

#rOCAT::install_packages() # you could ignore the line if you have 'OCAT', 'scikit-learn', and 'faiss' packages installed in your python env
```

### 2.1.2 import data

``` r
# load Test Zeisel data
data <- R.matlab::readMat('../data/Test_5_Zeisel.mat')

# extract the gene feature matrix and make it saved as a sparse matrix
in_X <- as(data$in.X, 'dgCMatrix')

# extract the labels
labels_true <- as.vector(data$true.labs)

# the input data should be a vector of datasets c(datasets1,dataset2,...)
data_list <- c(in_X)
```

### 2.1.3 run OCAT pipeline

``` r
ZW <- run_OCAT(data_list, m_list=list(50), dim=30, 
                p=0.3, log_norm=TRUE, l2_norm=TRUE, if_inference=FALSE, 
                random_seed=42)
```

### 2.1.4 clustering and evaluation

``` r
labels_pred <- evaluate_clusters(np_array(ZW),num_cluster = length(unique(labels_true)))

# get the NMI score between our predicted cell type and the true cell type 
normalized_mutual_info_score(labels_true, labels_pred)
```

## 2.2 Working with Multiple Datasets

### 2.2.1 import datasets

``` r
# the data is saved in .npz file in this example, so we need scipy from python to help read the data
scipy <- reticulate::import("scipy.sparse")
data_path <- "../data/Pancreas/data/"
files <- c('baron_1','muraro_2', 'seg_3', 'wang_4', 'xin_5')
data_list <- lapply(files,  FUN = function(x) scipy$load_npz(paste0(data_path,x,".npz")))

# import the labels
np <- reticulate::import("numpy") 
label_path <- "../data/Pancreas/label/"

# load the labels from each batch and combine them into one file
labels <- lapply(files,  FUN = function(x) np$load(paste0(label_path,x,"_label.npy"),allow_pickle = TRUE))
labels_true_combined <- do.call(c, labels)

# also save the batch number together
batch_true_combined <- rep(files,lengths(labels)) #batch labels 
```

### 2.2.2 run OCAT pipline

``` r
m_list <- list(65, 65, 65, 65, 65)

# execute the pipeline in these datasets
pancreatic_ZW <- run_OCAT(data_list, m_list, dim=60, p=0.3, log_norm=TRUE, l2_norm=TRUE)
```

### 2.2.3 clustering and evaluation

#### 2.2.3.1 cell type clustering

``` r
cell_type_pred <- evaluate_clusters(pancreatic_ZW, num_cluster = length(unique(labels_true_combined)))

# get the NMI score between our predicted cell type and the true cell type
normalized_mutual_info_score(labels_true_combined,cell_type_pred)
```

#### 2.2.3.2 batch clustering

``` r
batch_pred <- evaluate_clusters(pancreatic_ZW, num_cluster = length(files))

# get the NMI score between our predicted batch and the true batch
normalized_mutual_info_score(batch_true_combined,batch_pred)
```


NEWS.md
# rOCAT 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


R/utils.R

#' @description initialize required python packages
#' @export
install_packages <- function(){
  if (!reticulate::py_module_available("OCAT"))
    reticulate::py_install("git+https://github.com/bowang-lab/OCAT.git")
  if (!reticulate::py_module_available("faiss"))
    reticulate::py_install('faiss')
  if (!reticulate::py_module_available("sklearn"))
    reticulate::py_install("scikit-learn")
  if (!reticulate::py_module_available("umap"))
    reticulate::py_install("umap-learn")
  return(TRUE)
}

#' @description  normalize all cells by log10(x+1)
#' @export
normalize_data <- function(data_list){
  data_list <- sapply(data_list, function(x) log10(x+1))
  data_list <- sapply(data_list, function(x) Matrix::Matrix(x,sparse = TRUE))
  return(data_list)
}

#' @description  normalize all cells by L2 normalization
#' @export
l2_normalization <- function(data_list){
  for (i in seq_along(data_list)){
    l2_norm <- sqrt(Matrix::colSums(data_list[[i]]^2))
    l2_norm[l2_norm==0] <- 0.00001
    data_list[[i]] <- sweep(data_list[[i]],2,l2_norm,'/')
    data_list[[i]] <- Matrix::Matrix(data_list[[i]], sparse = TRUE)
  }
  return(data_list)
}

m_estimate <- function(data_list){
  m_list <- c()
  for (i in data_list){
    len <- i@Dim[1]
    if (len<2000)
      m_list <- append(m_list,20)
    else if (len<10000)
      m_list <- append(m_list,40)
    else
      m_list <- append(m_list,60)
  }
  print(paste0('estimated m_list: ', m_list))
  return(m_list)
}

#' @description  normalize cells expression
#' @param data_list   [(a,n)...(z,n)]    -- list of datasets.
#' @param log_norm  flag for log10 normalization.
#' @param l2_norm   flag  for L2 normalization.
#' Out:
#' @return data_list   [(a,dim)...(z,dim)]    -- list of normalized datasets
#'
#' @export
preprocess <- function(data_list, log_norm=log_norm, l2_norm=l2_norm){
  stopifnot("Data list cannot be empty"= length(data_list)>0)
  
  if (log_norm){
    data_list <- normalize_data(data_list)
  }
  if (l2_norm){
    data_list <- l2_normalization(data_list)
  }
  return(data_list)
}


dim_estimate <- function(data_list){
  gene_num <- data_list[[1]]@Dim[1]
  if (gene_num<5000)
    return(50)
  else if (gene_num<10000)
    return(100)
  else
    return(125)
}


#' @description  The function reduces the datasets to dim subspaces
#' @param data_list   [(a,n)...(z,n)]  list of datasets
#' @param dim   desired dimension after dimensionality reduction
#' @param mode  method for finding anchors, default on 'FSM'
#' @param upsample flag for upsample
#' @return data_list   [(a,dim)...(z,dim)] list of datasets
#' @export
apply_dim_reduct <- function(data_list, dim=NULL, mode='FSM', random_seed=42, upsample=FALSE){
  if (is.null(dim))
    dim <- dim_estimate(data_list)

  OCAT <- reticulate::import('OCAT')
  
  print("Starting Dimension Reduction")
  scipy <- reticulate::import("scipy.sparse",convert = FALSE)
  preprocessed_converted_data <- lapply(data_list, function(x) scipy$csr_matrix(reticulate::np_array(x,dtype='float32')))
  reticulate::py_run_string(glue::glue("dim = int({dim});random_seed = {random_seed};upsample={reticulate::r_to_py(upsample)};mode='{mode}'"))
  dim_reducted <- OCAT$apply_dim_reduct(preprocessed_converted_data,dim=py$dim,mode = py$mode, random_seed = py$random_seed,upsample = py$upsample)
  print("Dimension Reduction Finished")
  return(dim_reducted)
}

#' @description  The function generates the corresponding anchors for the datasets
#' @param data_list   [(a,n)...(z,n)] list of datasets
#' @param m_list   lift of num of anchors
#' Out:
#' @return anchor_list [(m,n)...(m,n)] list of anchors
#' @export
find_anchors <- function(data_list,m_list){
  faiss <- reticulate::import('faiss')
  
  anchor_list <- vector(mode = "list", length = length(data_list))
  for (X in seq_along(data_list)){
    dim1 <- dim(data_list[[X]])[[2]]
    temp <- reticulate::np_array(as(data_list[[X]],"matrix"),dtype = 'float32')
    m <- m_list[[X]]
    reticulate::py_run_string(glue::glue("dim={dim1};k={m}",convert=FALSE))
    kmean <- faiss$Kmeans(py$dim,py$k)
    kmean$train(temp)
    anchors <- kmean$centroids
    anchor_list[[X]] <- as(anchors,"matrix")
  }
  return(anchor_list)
}

#'@description  The function generates the sparsed encoding of edges
#' @param data_list   [(a,dim)...(z,dim)]  list of datasets (dim PCs)
#' @param m_list      num of anchors
#' @param s_list      num of anchors to be selected
#' @param p           percentage of NNs to consider
#' @param cn          rounds of optimization
#' @param if_inference     flag for cell inference
#' Out:
#' @return ZW        (a+...+z, m)  OCAT feature matrix
#'
#' @export
sparse_encoding_integration <- function(data_list, m_list, s_list, p=0.3, cn=5, if_inference=TRUE){
  OCAT <- reticulate::import('OCAT')
  anchor_list <- find_anchors(data_list, m_list)
  
  Z_list <- NULL
  for (i in seq_along(data_list)){
    dataset_Z_list <- NULL
    
    dataset <- data_list[[i]]
    dataset <- reticulate::np_array(dataset,dtype='float32')
    
    for (j in seq_along(anchor_list)){
      anchor <- reticulate::np_array(anchor_list[[j]],dtype='float32')
      reticulate::py_run_string(glue::glue("s = {s_list[[j]]};flag=2;cn = {cn}"))
      Z <- OCAT$sparse_encoding$AnchorGraph(dataset$transpose(),anchor$transpose(),py$s,py$flag,py$cn)
      Z <- as(Z,'matrix')
      dataset_Z_list <- cbind(dataset_Z_list,Z)
      dataset_Z_list <- as(dataset_Z_list,'matrix')
    }
    Z_list <- rbind(Z_list,dataset_Z_list)
    dataset_Z_list <- as(Z_list,'matrix') #make sure it keeps the dim with multiple datasets
  }
  Z <- reticulate::np_array(Z_list)
  ZW_Wanchor <- OCAT$sparse_encoding$Z_to_ZW(Z)
  ZW <- ZW_Wanchor[[1]]
  ZW[is.na(ZW)] <- 0
  W_anchor <- ZW_Wanchor[[2]]
  ZW <- OCAT$sparse_encoding$norm(ZW)
  if (if_inference)
    return(list(ZW,anchor_list,s_list,W_anchor))
  else
    return(ZW)
}


#' @description  The function performs OCAT pipeline and return the sparsed encoding oof edges
#' @param object  Seurat Object
#' @param m_list  num of anchors
#' @param s_list  num of anchors to be selected
#' @param dim     default values of dim
#' @param p       percentage of NNs to consider,default = 0.3
#' @param log_norm  num of anchors to be selected
#' @param if_inference  flag for cell inference
#' @param random_seed default = 42
#' Out:
#' @return ZW (a+...+z, m)  OCAT feature matrix
#' @export
run_OCAT <- function(object, group.by.vars, ...) {
  UseMethod("run_OCAT")
}


run_OCAT.default <- function(data_list, m_list=NULL, s_list=NULL, dim=NULL,
                     p=0.3, log_norm=TRUE, l2_norm=TRUE, if_inference=FALSE,
                     random_seed=42){
  if (is.null(m_list))
    m_list <- m_estimate(data_list)
  if (is.null(s_list))
    s_list <- lapply(m_list,"*",p)
    s_list <- lapply(s_list,round)
  if (is.null(dim))
      dim <- dim_estimate(data_list)
  
  data_list <- preprocess(data_list, log_norm=TRUE, l2_norm=TRUE)
  Wm <- NULL
  
  if (if_inference){
    data_list_Wm <- apply_dim_reduct(data_list, dim=dim, mode='FSM', random_seed=random_seed)
    data_list <- data_list_Wm[[1]]
    Wm <- data_list_Wm[[2]]
    sparse_list <- sparse_encoding_integration(data_list, m_list=m_list, s_list=s_list, p=p, cn=5, if_inference=TRUE)
    db_list <- list(sparse_list[[2]],s_list,sparse_list[[4]],Wm)
    return(list(ZW,db_list))
  }
  else {
    data_list <- apply_dim_reduct(data_list, dim=dim, mode='FSM', random_seed=random_seed)[[1]]
    ZW <- sparse_encoding_integration(data_list, m_list=m_list, s_list=s_list, p=p, cn=5, if_inference=FALSE)
    return(ZW)
  }
}



run_OCAT.Seurat <- function(object, reduction.name ='ocat',m_list=NULL, s_list=NULL, dim=NULL,
                             p=0.3, log_norm=TRUE, l2_norm=TRUE, if_inference=FALSE,
                             random_seed=42, assay = "RNA"){
  
  data_list <- c(t(object@assays$RNA@data))
  ZW <- run_OCAT.Default(data_list, m_list, s_list, dim,
                         p, log_norm, l2_norm, if_inference,
                         random_seed) 
  rownames(ZW) <- rownames(data_list[[1]])
  colnames(ZW) <- paste0(reduction.name, "_", seq_len(ncol(ZW)))
  suppressWarnings({
    ocat_data <- Seurat::CreateDimReducObject(embeddings = as(ZW,"matrix"), 
                                                stdev = as.numeric(apply(ZW, 2, stats::sd)), 
                                                assay = assay, key = reduction.name)
  })
  object[[reduction.name]] <- ocat_data
  return(object)
}


#' @description  The function wraps the evaluate_cluster function from OCAT python packagee
#' @param ZW   OCAT feature matrix
#' @param num_clusters  num of clusters
#' Out:
#' @return labels_pred  predicted labels 
#' @export
evaluate_clusters <- function(ZW, num_clusters){
  OCAT <- reticulate::import('OCAT')
  labels_pred <- OCAT$evaluate_clusters(reticulate::np_array(ZW),num_clusters)
  return(labels_pred)
}

#' @description The function wraps the normalized_mutual_info_score from sklearn.metrics.cluster
#' @param labels_true  ground truth labels
#' @param labels_pred  predicted labels
#' Out:
#' @return NMI     NMI score
#' @export
normalized_mutual_info_score <- function(labels_true, labels_pred){
  sklearn <- reticulate::import('sklearn')
  NMI <- sklearn$metrics$cluster$normalized_mutual_info_score(labels_true, labels_pred)
  return(NMI)
}

R/visual_utils.R
#' @title run_UMAP
#' @description  The function calculate the UMAP embedding
#' @param embeddding  OCAT embedding
#' Out:
#' @return OCAT_umap calculated UMAP embedding
#' @export
run_UMAP <- function(embeddding){
  if (!reticulate::py_module_available("umap"))
    reticulate::py_install("umap-learn")
  umap <- reticulate::import('umap')
  OCAT_umap <- umap$UMAP()$fit_transform(embeddding)
  return(OCAT_umap)
}

gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}

#' @title plot_UMAP
#' @description  The function plot the UMAP embedding
#' @param embeddding  UMAP embedding
#' @param labels_true  UMAP embedding
#' @param labels_predicted  UMAP embedding
#' @param legend_labels  UMAP embedding
#' @param title  UMAP embedding
#' @param x  the x and y co-ordinates to be used to position the legend. They can be specified by keyword or in any way which is accepted by xy.coords
#' @param y.intersp  character interspacing factor for vertical (y) line distance
#' @export
plot_UMAP <- function(umap_embedding,labels_true,labels_predicted, legend_labels = NULL, title = '', x='bottomright', y.intersp=1){
  unique_cell_types <- sort(unique(labels_predicted))
  cell_type_color <- gg_color_hue(length(unique_cell_types))

  if (is.null(legend_labels))
    legend_labels <- unique_cell_types

  plot(umap_embedding[,1], umap_embedding[,2], type='n', xlab='UMAP_1', ylab='UMAP_2',
       yaxt  ='n',xaxt='n',
       bty='L', tck=FALSE,fg='grey', mgp=c(1,1,0), main=title)

  for(i in 1:length(unique_cell_types)){
    temp_ind <- which(labels_predicted==unique_cell_types[i])
    points(umap_embedding[temp_ind,1], umap_embedding[temp_ind,2],
           col=cell_type_color[i], pch=20, cex=0.5)
  }
  legend(x,
         legend=legend_labels,
         col=cell_type_color, bty='n', pch=19, y.intersp=y.intersp)
}


R/oxy-utils.R

#' @title FUNCTION_TITLE
#' @description FUNCTION_DESCRIPTION
#' @return OUTPUT_DESCRIPTION
#' @details DETAILS
#' @examples 
#' \dontrun{
#' if(interactive()){
#'  #EXAMPLE1
#'  }
#' }
#' @seealso 
#'  \code{\link[reticulate]{py_module_available}}, \code{\link[reticulate]{py_install}}
#' @rdname install_packages
#' @export 
#' @importFrom reticulate py_module_available py_install
install_packages <- function(){
  if (!reticulate::py_module_available("OCAT"))
    reticulate::py_install("git+https://github.com/bowang-lab/OCAT.git")
  if (!reticulate::py_module_available("faiss"))
    reticulate::py_install('faiss')
  if (!reticulate::py_module_available("sklearn"))
    reticulate::py_install("scikit-learn")
  return(TRUE)
}

#' @title FUNCTION_TITLE
#' @description FUNCTION_DESCRIPTION
#' @param data_list PARAM_DESCRIPTION
#' @return OUTPUT_DESCRIPTION
#' @details DETAILS
#' @examples 
#' \dontrun{
#' if(interactive()){
#'  #EXAMPLE1
#'  }
#' }
#' @seealso 
#'  \code{\link[Matrix]{Matrix}}
#' @rdname normalize_data
#' @export 
#' @importFrom Matrix Matrix
normalize_data <- function(data_list){
  data_list <- sapply(data_list, function(x) log10(x+1))
  data_list <- sapply(data_list, function(x) Matrix::Matrix(x,sparse = TRUE))
  return(data_list)
}

#' @title FUNCTION_TITLE
#' @description FUNCTION_DESCRIPTION
#' @param data_list PARAM_DESCRIPTION
#' @return OUTPUT_DESCRIPTION
#' @details DETAILS
#' @examples 
#' \dontrun{
#' if(interactive()){
#'  #EXAMPLE1
#'  }
#' }
#' @seealso 
#'  \code{\link[Matrix]{colSums}}, \code{\link[Matrix]{Matrix}}
#' @rdname l2_normalization
#' @export 
#' @importFrom Matrix colSums Matrix
l2_normalization <- function(data_list){
  for (i in seq_along(data_list)){
    l2_norm <- sqrt(Matrix::colSums(data_list[[i]]^2))
    l2_norm[l2_norm==0] <- 0.00001
    data_list[[i]] <- sweep(data_list[[i]],2,l2_norm,'/')
    data_list[[i]] <- Matrix::Matrix(data_list[[i]], sparse = TRUE)
  }
  return(data_list)
}

#' @title FUNCTION_TITLE
#' @description FUNCTION_DESCRIPTION
#' @param data_list PARAM_DESCRIPTION
#' @return OUTPUT_DESCRIPTION
#' @details DETAILS
#' @examples 
#' \dontrun{
#' if(interactive()){
#'  #EXAMPLE1
#'  }
#' }
#' @rdname m_estimate
#' @export 
m_estimate <- function(data_list){
  m_list <- c()
  for (i in data_list){
    len <- i@Dim[1]
    if (len<2000)
      m_list <- append(m_list,20)
    else if (len<10000)
      m_list <- append(m_list,40)
    else
      m_list <- append(m_list,60)
  }
  print(paste0('estimated m_list: ', m_list))
  return(m_list)
}

#' @title FUNCTION_TITLE
#' @description FUNCTION_DESCRIPTION
#' @param data_list PARAM_DESCRIPTION
#' @param log_norm PARAM_DESCRIPTION, Default: log_norm
#' @param l2_norm PARAM_DESCRIPTION, Default: l2_norm
#' @return OUTPUT_DESCRIPTION
#' @details DETAILS
#' @examples 
#' \dontrun{
#' if(interactive()){
#'  #EXAMPLE1
#'  }
#' }
#' @rdname preprocess
#' @export 
preprocess <- function(data_list, log_norm=log_norm, l2_norm=l2_norm){
  stopifnot("Data list cannot be empty"= length(data_list)>0)
  
  if (log_norm){
    data_list <- normalize_data(data_list)
  }
  if (l2_norm){
    data_list <- l2_normalization(data_list)
  }
  return(data_list)
}


#' @title FUNCTION_TITLE
#' @description FUNCTION_DESCRIPTION
#' @param data_list PARAM_DESCRIPTION
#' @return OUTPUT_DESCRIPTION
#' @details DETAILS
#' @examples 
#' \dontrun{
#' if(interactive()){
#'  #EXAMPLE1
#'  }
#' }
#' @rdname dim_estimate
#' @export 
dim_estimate <- function(data_list){
  gene_num <- data_list[[1]]@Dim[1]
  if (gene_num<5000)
    return(50)
  else if (gene_num<10000)
    return(100)
  else
    return(125)
}


#' @title FUNCTION_TITLE
#' @description FUNCTION_DESCRIPTION
#' @param data_list PARAM_DESCRIPTION
#' @param dim PARAM_DESCRIPTION, Default: NULL
#' @param mode PARAM_DESCRIPTION, Default: 'FSM'
#' @param random_seed PARAM_DESCRIPTION, Default: 42
#' @param upsample PARAM_DESCRIPTION, Default: FALSE
#' @return OUTPUT_DESCRIPTION
#' @details DETAILS
#' @examples 
#' \dontrun{
#' if(interactive()){
#'  #EXAMPLE1
#'  }
#' }
#' @seealso 
#'  \code{\link[reticulate]{import}}, \code{\link[reticulate]{np_array}}, \code{\link[reticulate]{py_run}}
#'  \code{\link[glue]{glue}}
#' @rdname apply_dim_reduct
#' @export 
#' @importFrom reticulate import np_array py_run_string
#' @importFrom glue glue
apply_dim_reduct <- function(data_list, dim=NULL, mode='FSM', random_seed=42, upsample=FALSE){
  if (is.null(dim))
    dim <- dim_estimate(data_list)

  OCAT <- reticulate::import('OCAT')
  
  print("Starting Dimension Reduction")
  scipy <- reticulate::import("scipy.sparse",convert = FALSE)
  preprocessed_converted_data <- lapply(data_list, function(x) scipy$csr_matrix(reticulate::np_array(x,dtype='float32')))
  reticulate::py_run_string(glue::glue("dim = int({dim});random_seed = {random_seed};upsample={reticulate::r_to_py(upsample)};mode='{mode}'"))
  dim_reducted <- OCAT$apply_dim_reduct(preprocessed_converted_data,dim=py$dim,mode = py$mode, random_seed = py$random_seed,upsample = py$upsample)
  print("Dimension Reduction Finished")
  return(dim_reducted)
}

#' @title FUNCTION_TITLE
#' @description FUNCTION_DESCRIPTION
#' @param data_list PARAM_DESCRIPTION
#' @param m_list PARAM_DESCRIPTION
#' @return OUTPUT_DESCRIPTION
#' @details DETAILS
#' @examples 
#' \dontrun{
#' if(interactive()){
#'  #EXAMPLE1
#'  }
#' }
#' @seealso 
#'  \code{\link[reticulate]{import}}, \code{\link[reticulate]{np_array}}, \code{\link[reticulate]{py_run}}
#'  \code{\link[glue]{glue}}
#' @rdname find_anchors
#' @export 
#' @importFrom reticulate import np_array py_run_string
#' @importFrom glue glue
find_anchors <- function(data_list,m_list){
  faiss <- reticulate::import('faiss')
  
  anchor_list <- vector(mode = "list", length = length(data_list))
  for (X in seq_along(data_list)){
    dim1 <- dim(data_list[[X]])[[2]]
    temp <- reticulate::np_array(as(data_list[[X]],"matrix"),dtype = 'float32')
    m <- m_list[[X]]
    reticulate::py_run_string(glue::glue("dim={dim1};k={m}",convert=FALSE))
    kmean <- faiss$Kmeans(py$dim,py$k)
    kmean$train(temp)
    anchors <- kmean$centroids
    anchor_list[[X]] <- as(anchors,"matrix")
  }
  return(anchor_list)
}

#' @title FUNCTION_TITLE
#' @description FUNCTION_DESCRIPTION
#' @param data_list PARAM_DESCRIPTION
#' @param m_list PARAM_DESCRIPTION
#' @param s_list PARAM_DESCRIPTION
#' @param p PARAM_DESCRIPTION, Default: 0.3
#' @param cn PARAM_DESCRIPTION, Default: 5
#' @param if_inference PARAM_DESCRIPTION, Default: TRUE
#' @return OUTPUT_DESCRIPTION
#' @details DETAILS
#' @examples 
#' \dontrun{
#' if(interactive()){
#'  #EXAMPLE1
#'  }
#' }
#' @seealso 
#'  \code{\link[reticulate]{import}}, \code{\link[reticulate]{np_array}}, \code{\link[reticulate]{py_run}}
#'  \code{\link[glue]{glue}}
#' @rdname sparse_encoding_integration
#' @export 
#' @importFrom reticulate import np_array py_run_string
#' @importFrom glue glue
sparse_encoding_integration <- function(data_list, m_list, s_list, p=0.3, cn=5, if_inference=TRUE){
  OCAT <- reticulate::import('OCAT')
  anchor_list <- find_anchors(data_list, m_list)
  
  Z_list <- NULL
  for (i in seq_along(data_list)){
    dataset_Z_list <- NULL
    
    dataset <- data_list[[i]]
    dataset <- reticulate::np_array(dataset,dtype='float32')
    
    for (j in seq_along(anchor_list)){
      anchor <- reticulate::np_array(anchor_list[[j]],dtype='float32')
      reticulate::py_run_string(glue::glue("s = {s_list[[j]]};flag=2;cn = {cn}"))
      Z <- OCAT$sparse_encoding$AnchorGraph(dataset$transpose(),anchor$transpose(),py$s,py$flag,py$cn)
      Z <- as(Z,'matrix')
      dataset_Z_list <- cbind(dataset_Z_list,Z)
      dataset_Z_list <- as(dataset_Z_list,'matrix')
    }
    Z_list <- rbind(Z_list,dataset_Z_list)
    dataset_Z_list <- as(Z_list,'matrix') #make sure it keeps the dim with multiple datasets
  }
  Z <- reticulate::np_array(Z_list)
  ZW_Wanchor <- OCAT$sparse_encoding$Z_to_ZW(Z)
  ZW <- ZW_Wanchor[[1]]
  ZW[is.na(ZW)] <- 0
  W_anchor <- ZW_Wanchor[[2]]
  ZW <- OCAT$sparse_encoding$norm(ZW)
  if (if_inference)
    return(list(ZW,anchor_list,s_list,W_anchor))
  else
    return(ZW)
}


#' @title FUNCTION_TITLE
#' @description FUNCTION_DESCRIPTION
#' @param object PARAM_DESCRIPTION
#' @param group.by.vars PARAM_DESCRIPTION
#' @param ... PARAM_DESCRIPTION
#' @return OUTPUT_DESCRIPTION
#' @details DETAILS
#' @examples 
#' \dontrun{
#' if(interactive()){
#'  #EXAMPLE1
#'  }
#' }
#' @rdname run_OCAT
#' @export 
run_OCAT <- function(object, group.by.vars, ...) {
  UseMethod("run_OCAT")
}


#' @title FUNCTION_TITLE
#' @description FUNCTION_DESCRIPTION
#' @param data_list PARAM_DESCRIPTION
#' @param m_list PARAM_DESCRIPTION, Default: NULL
#' @param s_list PARAM_DESCRIPTION, Default: NULL
#' @param dim PARAM_DESCRIPTION, Default: NULL
#' @param p PARAM_DESCRIPTION, Default: 0.3
#' @param log_norm PARAM_DESCRIPTION, Default: TRUE
#' @param l2_norm PARAM_DESCRIPTION, Default: TRUE
#' @param if_inference PARAM_DESCRIPTION, Default: FALSE
#' @param random_seed PARAM_DESCRIPTION, Default: 42
#' @return OUTPUT_DESCRIPTION
#' @details DETAILS
#' @examples 
#' \dontrun{
#' if(interactive()){
#'  #EXAMPLE1
#'  }
#' }
#' @rdname run_OCAT.default
#' @export 
run_OCAT.default <- function(data_list, m_list=NULL, s_list=NULL, dim=NULL,
                     p=0.3, log_norm=TRUE, l2_norm=TRUE, if_inference=FALSE,
                     random_seed=42){
  if (is.null(m_list))
    m_list <- m_estimate(data_list)
  if (is.null(s_list))
    s_list <- lapply(m_list,"*",p)
    s_list <- lapply(s_list,round)
  if (is.null(dim))
      dim <- dim_estimate(data_list)
  
  data_list <- preprocess(data_list, log_norm=TRUE, l2_norm=TRUE)
  Wm <- NULL
  
  if (if_inference){
    data_list_Wm <- apply_dim_reduct(data_list, dim=dim, mode='FSM', random_seed=random_seed)
    data_list <- data_list_Wm[[1]]
    Wm <- data_list_Wm[[2]]
    sparse_list <- sparse_encoding_integration(data_list, m_list=m_list, s_list=s_list, p=p, cn=5, if_inference=TRUE)
    db_list <- list(sparse_list[[2]],s_list,sparse_list[[4]],Wm)
    return(list(ZW,db_list))
  }
  else {
    data_list <- apply_dim_reduct(data_list, dim=dim, mode='FSM', random_seed=random_seed)[[1]]
    ZW <- sparse_encoding_integration(data_list, m_list=m_list, s_list=s_list, p=p, cn=5, if_inference=FALSE)
    return(ZW)
  }
}



#' @title FUNCTION_TITLE
#' @description FUNCTION_DESCRIPTION
#' @param object PARAM_DESCRIPTION
#' @param reduction.name PARAM_DESCRIPTION, Default: 'ocat'
#' @param m_list PARAM_DESCRIPTION, Default: NULL
#' @param s_list PARAM_DESCRIPTION, Default: NULL
#' @param dim PARAM_DESCRIPTION, Default: NULL
#' @param p PARAM_DESCRIPTION, Default: 0.3
#' @param log_norm PARAM_DESCRIPTION, Default: TRUE
#' @param l2_norm PARAM_DESCRIPTION, Default: TRUE
#' @param if_inference PARAM_DESCRIPTION, Default: FALSE
#' @param random_seed PARAM_DESCRIPTION, Default: 42
#' @param assay PARAM_DESCRIPTION, Default: 'RNA'
#' @return OUTPUT_DESCRIPTION
#' @details DETAILS
#' @examples 
#' \dontrun{
#' if(interactive()){
#'  #EXAMPLE1
#'  }
#' }
#' @seealso 
#'  \code{\link[Seurat]{reexports}}
#'  \code{\link[stats]{sd}}
#' @rdname run_OCAT.Seurat
#' @export 
#' @importFrom Seurat CreateDimReducObject
#' @importFrom stats sd
run_OCAT.Seurat <- function(object, reduction.name ='ocat',m_list=NULL, s_list=NULL, dim=NULL,
                             p=0.3, log_norm=TRUE, l2_norm=TRUE, if_inference=FALSE,
                             random_seed=42, assay = "RNA"){
  
  data_list <- c(t(object@assays$RNA@data))
  ZW <- run_OCAT.Default(data_list, m_list, s_list, dim,
                         p, log_norm, l2_norm, if_inference,
                         random_seed) 
  rownames(ZW) <- rownames(data_list[[1]])
  colnames(ZW) <- paste0(reduction.name, "_", seq_len(ncol(ZW)))
  suppressWarnings({
    ocat_data <- Seurat::CreateDimReducObject(embeddings = as(ZW,"matrix"), 
                                                stdev = as.numeric(apply(ZW, 2, stats::sd)), 
                                                assay = assay, key = reduction.name)
  })
  object[[reduction.name]] <- ocat_data
  return(object)
}


#' @title FUNCTION_TITLE
#' @description FUNCTION_DESCRIPTION
#' @param ZW PARAM_DESCRIPTION
#' @param num_clusters PARAM_DESCRIPTION
#' @return OUTPUT_DESCRIPTION
#' @details DETAILS
#' @examples 
#' \dontrun{
#' if(interactive()){
#'  #EXAMPLE1
#'  }
#' }
#' @seealso 
#'  \code{\link[reticulate]{import}}, \code{\link[reticulate]{np_array}}
#' @rdname evaluate_clusters
#' @export 
#' @importFrom reticulate import np_array
evaluate_clusters <- function(ZW, num_clusters){
  OCAT <- reticulate::import('OCAT')
  labels_pred <- OCAT$evaluate_clusters(reticulate::np_array(ZW),num_clusters)
  return(labels_pred)
}

#' @title FUNCTION_TITLE
#' @description FUNCTION_DESCRIPTION
#' @param labels_true PARAM_DESCRIPTION
#' @param labels_pred PARAM_DESCRIPTION
#' @return OUTPUT_DESCRIPTION
#' @details DETAILS
#' @examples 
#' \dontrun{
#' if(interactive()){
#'  #EXAMPLE1
#'  }
#' }
#' @seealso 
#'  \code{\link[reticulate]{import}}
#' @rdname normalized_mutual_info_score
#' @export 
#' @importFrom reticulate import
normalized_mutual_info_score <- function(labels_true, labels_pred){
  sklearn <- reticulate::import('sklearn')
  NMI <- sklearn$metrics$cluster$normalized_mutual_info_score(labels_true, labels_pred)
  return(NMI)
}
